feat: client credentials

sdelamo opened the pull request:
close: https://github.com/micronaut-projects/micronaut-security/issues/12

sdelamo commented:
@graemerocher @jameskleeh could you check this PR. Specially the way I am caching the publisher result. See `CacheablePublisher`. Not sure if there is a better way to do this.

jameskleeh commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/ClientCredentialsHttpClientFilter.java:
```diff
@@ -0,0 +1,170 @@
+    protected final Collection<OauthClientConfiguration> oauthClientConfigurationCollection;
+    protected final TokenPropagator tokenPropagator;
+    protected final BeanContext beanContext;
+    protected final Map<String, ClientCredentialsClient> clientCredentialsClientsByName = new ConcurrentHashMap<>();
```
This needs to be thread safe

sdelamo replied:
```java
if (!clientCredentialsClientsByName.containsKey(oauthClient.getName())) {
    try {
        ClientCredentialsClient clientCredentialsClient = beanContext.getBean(ClientCredentialsClient.class, Qualifiers.byName(oauthClient.getName()));
        clientCredentialsClientsByName.put(oauthClient.getName(), clientCredentialsClient);
    } catch (NoSuchBeanException e) {
        if (LOG.isTraceEnabled()) {
            LOG.trace("no client credentials client for OAuth 2.0 client {}", oauthClient.getName());
        }
    }
}
return Optional.ofNullable(clientCredentialsClientsByName.get(oauthClient.getName()));

```

with:

```java
try {
    clientCredentialsClientsByName.putIfAbsent(oauthClient.getName(), beanContext.getBean(ClientCredentialsClient.class, Qualifiers.byName(oauthClient.getName())));
    return Optional.of(clientCredentialsClientsByName.get(oauthClient.getName()));
} catch (NoSuchBeanException e) {
    if (LOG.isTraceEnabled()) {
        LOG.trace("no client credentials client for OAuth 2.0 client {}", oauthClient.getName());
    }
}
return Optional.empty();

```

jameskleeh replied:
That is still not good because that means the `getBean` call will happen every time and multiple puts could occur

sdelamo replied:
I've replaced the usage of `putIfAbsent` with `computeIfAbsent`

jameskleeh commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfiguration.java:
```diff
@@ -83,6 +84,13 @@
+     * @return The Client Credentials Configuration
+     */
+    @NonNull
+    Optional<ClientCredentialsConfiguration> getClientCredentials();
```
This needs a default implementation to not be a breaking change

sdelamo replied:
I've added a default implementation which returns an empty Optional.

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/OpenIdClientFactory.java:
```diff
@@ -90,7 +90,11 @@ DefaultOpenIdProviderMetadata openIdConfiguration(@Parameter OauthClientConfigur
                         }
-                        return issuerClient.toBlocking().retrieve(configurationUrl.toString(), DefaultOpenIdProviderMetadata.class);
+                        DefaultOpenIdProviderMetadata openIdProviderMetadata = issuerClient.toBlocking().retrieve(configurationUrl.toString(), DefaultOpenIdProviderMetadata.class);
+                        if (LOG.isTraceEnabled()) {
```
I've removed the log statement. It is not client credentials related. We can add it in a separate PR.

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/AbstractClientCredentialsClient.java:
```diff
@@ -0,0 +1,127 @@
+                LOG.trace("cannot parse access token {} to JWT", tokenResponse.getAccessToken());
+            }
+        }
+        if (tokenResponse instanceof TokenResponseExpiration) {
+            TokenResponseExpiration tokenResponseExpiration = (TokenResponseExpiration) tokenResponse;
+            return Optional.ofNullable(tokenResponseExpiration.getExpiration());
+        }
```
I think the Json Web Token expiration date claim is more precise than the `expiresIn` value. 

The `expiresIn` is the time to live in seconds for the token since the authorization server issued it. However, there is a network latency involved. I cannot really know when the token was issued. Thus, the expiration date calculated with the `expiresIn` is more blurry. 

Exp claim is an epoch value. Thus, I think it is slightly more precise.

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/AbstractClientCredentialsClient.java:
```diff
@@ -0,0 +1,127 @@
+    @NonNull
+    public Publisher<TokenResponse> clientCredentials(@Nullable String scope, boolean force) {
+        String resolvedScope = scope != null ? scope : NOSCOPE;
+        if (!scopeToPublisherMap.containsKey(resolvedScope)) {
+            scopeToPublisherMap.put(resolvedScope, new CacheablePublisher<>(TokenResponseExpiration::new));
+        }
+        CacheablePublisher<TokenResponse> publisher = scopeToPublisherMap.get(resolvedScope);
```
Yes, it is supposed to be thread safe.

sdelamo replied:
I’ve replaced: 

```java
if (!scopeToPublisherMap.containsKey(resolvedScope)) {
    scopeToPublisherMap.put(resolvedScope, new CacheableProcessor<>(TokenResponseExpiration::new));
}
```

with:

```java
scopeToPublisherMap.putIfAbsent(resolvedScope, new CacheableProcessor<>(TokenResponseExpiration::new));
```

sdelamo replied:
I've replaced the usage of `putIfAbsent` with `computeIfAbsent`

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/CacheablePublisher.java:
```diff
@@ -0,0 +1,259 @@
+ * @param <T> The element to be returned by the publisher
+ */
+@Experimental
+public class CacheablePublisher<T> implements Subscriber<T>, Publisher<T>, ElementsRequestedListener {
```
I've renamed it to `CacheableProcessor`.

graemerocher replied:
Seems to me you have just rewritten http://reactivex.io/RxJava/javadoc/io/reactivex/processors/BehaviorProcessor.html or am I missing something? Why the need for a custom processor impl?

sdelamo replied:
I was not aware of `BehaviorProcessor`

sdelamo replied:
I've replaced the usage of `putIfAbsent` with `computeIfAbsent`

sdelamo commented:
I have added a `advanced-expiration` settings to client credentials configuration. It defaults to 30 seconds. Thus, thirty seconds prior the calculated expiration date, tokens are considered expired. 

To have a default value even if the user does not specify anything in configuration I needed to set: 


```java
public class OauthClientConfigurationProperties {
... 
..
.
@NonNull
public Optional<ClientCredentialsConfiguration> getClientCredentials() {
    return Optional.of(clientCredentials == null ? new ClientCredentialsConfigurationProperties() : clientCredentials);
}
...
..
.
}
```

@jameskleeh  do you think I should just change the signature and do:


```java
public class OauthClientConfigurationProperties {
... 
..
.
@NonNull
public ClientCredentialsConfiguration getClientCredentials() {
    return clientCredentials == null ? new ClientCredentialsConfigurationProperties() : clientCredentials;
}
...
..
.
}
```

and change the interface to

```java
@NonNull
default ClientCredentialsConfiguration getClientCredentials() {
    return new OauthClientConfigurationProperties.ClientCredentialsConfigurationProperties();
}
```

Other nested configuration `OauthClientConfiguration` return `Optional`.

ttzn commented:
This feature would be very much appreciated. Maybe draw inspiration from the [spring-security implementation](https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ClientCredentialsOAuth2AuthorizedClientProvider.java) (which was recently rewritten).

sdelamo commented:
> I have added a `advanced-expiration` settings to client credentials configuration. It defaults to 30 seconds. Thus, thirty seconds prior the calculated expiration date, tokens are considered expired.
> 
> To have a default value even if the user does not specify anything in configuration I needed to set:
> 
> ```java
> public class OauthClientConfigurationProperties {
> ... 
> ..
> .
> @NonNull
> public Optional<ClientCredentialsConfiguration> getClientCredentials() {
>     return Optional.of(clientCredentials == null ? new ClientCredentialsConfigurationProperties() : clientCredentials);
> }
> ...
> ..
> .
> }
> ```
> 
> @jameskleeh do you think I should just change the signature and do:
> 
> ```java
> public class OauthClientConfigurationProperties {
> ... 
> ..
> .
> @NonNull
> public ClientCredentialsConfiguration getClientCredentials() {
>     return clientCredentials == null ? new ClientCredentialsConfigurationProperties() : clientCredentials;
> }
> ...
> ..
> .
> }
> ```
> 
> and change the interface to
> 
> ```java
> @NonNull
> default ClientCredentialsConfiguration getClientCredentials() {
>     return new OauthClientConfigurationProperties.ClientCredentialsConfigurationProperties();
> }
> ```
> 
> Other nested configuration in `OauthClientConfiguration` return `Optional`.

I've moved the default to the interface. If the user does not specify any setting for `client-credentials` that default value is used even if `getClientCredentials()` returns an empty optional.
